### PR TITLE
Extend du-config.json to support custom folders

### PR DESCRIPTION
**Pull request by Nox-MSFT**
#  Enabling Customer Overrides for Default Paths
>Note | Internal tracking task #43366149 & #43150953

This pull request enables customers to customize default paths by specifying a custom path in the du-config.json file.

The following build time configurations (compiler directives) can be overridden by specifying desired values in the du-config.json file:

| Variable | du-config property name | Default value | Override value |
|---|---|---|---|
| ADUC_DATA_FOLDER | dataFolder | /var/lib/adu | {dataFolder} |
| ADUC_DOWNLOADS_FOLDER | downloadsFolder | /var/lib/adu/downloads | {downloadsFolder} <br/> or fallback to {dataFolder}/downloads <br/> or fallback to default value |
| ADUC_EXTENSIONS_FOLDER | extensionsFolder | /var/lib/adu/extensions | {extensionsFolder} <br/>or fallback to {dataFolder}/extensions <br/> or fallback to default value |
| ADUC_COMPONENT_ENUMERATOR_EXTENSION_DIR | N/A | /var/lib/adu/extensions/component_enumerator | {extensionsFolder}/component_enumerator |
| ADUC_UPDATE_CONTENT_HANDLER_EXTENSION_DIR | N/A | /var/lib/adu/extensions/update_content_handlers | {extensionsFolder}/update_content_handlers |
| ADUC_CONTENT_DOWNLOADER_EXTENSION_DIR | N/A | /var/lib/adu/extensions/content_downloader | {extensionsFolder}/content_downloader |
| ADUC_DOWNLOAD_HANDLER_EXTENSION_DIR | N/A | /var/lib/adu/extensions/download_handlers | {extensionsFolder}/download_handlers |
| ADUSHELL_FOLDER | aduShellFolder | /usr/lib/adu | {aduShellFolder} |
| ADUSHELL_FILE_PATH | N/A | /usr/lib/adu/adu-shell | {aduShellFolder}/adu-shell |

## Specifying a custom du-config.json path
You can specify a custom du-config.json folder by using the aducIotAgent command-line options "-F" or "--config-folder".

For example:
```sh
# Load du-config.json from /user/local/adu/du-config.json
# instead of the default: /etc/adu/du-config.json
aducIotAgent -F "/user/local/adu"
```

To ensure that adu-shell also loads configurations from the same location, the same du-config folder will be passed to adu-shell at runtime.

By default, adu-shell will be invoked with "-F /etc/adu". If aducIotAgent is invoked with "-F /user/local/adu", this path value will be passed along to adu-shell.

## Change adu-shell default path
This pull request also change the adu-shell path from /usr/lib/adu to /usr/bin

### CODE REVIEW

File: src/adu-shell/src/main.cpp [hunk 1]
```diff
@@ -350,9 +374,14 @@ int main(int argc, char** argv)
...
 
-        return ret;
+        goto done;
     }
 
     Log_Error("Cannot set user identity. (code: %d, errno: %d)", ret, errno);
+done:
+    if (config != NULL)
+    {
+        ADUC_ConfigInfo_ReleaseInstance(config);
```
Comment (nihemstr) [inline src/adu-shell/src/main.cpp:384]: nit: there's a few places here where we do a NULL check before release. Might be good to make the ReleaseInstance() function handle NULL so we have an easier escape path all around.

File: src/adu_workflow/src/agent_workflow.c [hunk 2]
```diff
@@ -555,7 +556,17 @@ void ADUC_Workflow_HandlePropertyUpdate(
     else
     {
         // This is a top level workflow, make sure that we set the working folder correctly.
-        workflow_set_workfolder(nextWorkflow, "%s/%s", ADUC_DOWNLOADS_FOLDER, workflow_peek_id(nextWorkflow));
+        const ADUC_ConfigInfo* config = ADUC_ConfigInfo_GetInstance();
+        if (config != NULL)
```
Comment (nihemstr) [inline src/adu_workflow/src/agent_workflow.c:560]: nit: would be clean to add a ConfigIsNotNUllOrEmpty() function to make these checks more aggressive and cleaner.

File: src/utils/config_utils/src/config_utils.c [hunk 3]
```diff
@@ -15,11 +15,48 @@
...
+    pthread_mutex_lock(&s_config_mutex);
+}
+
+static inline void s_config_unlock(void)
+{
+    pthread_mutex_unlock(&s_config_mutex);
+}
+
+//static const char* CONFIG_LOG_FOLDER = "logFolder";
+static const char* CONFIG_ADU_SHELL_FOLDER = "aduShellFolder";
```
Comment (nihemstr) [inline src/utils/config_utils/src/config_utils.c:39]: Do these need to be static? We could make these #defines at the top for more readability/utility. Probably a nit but wanted to suggest. 

Also commented out CONFIG_LOG_FOLDER probably needs to be removed.

**Review by nihemstr — APPROVED** [review]
Approved with some nits.

**Comment by Nox-MSFT** [comment]
Found some issues with this changes. Will close this PR and create a new one.

**Pull request closed without merging**